### ISSUE TITLE
Link to known attestations in repository

### ISSUE THREAD

**Issue by joshuagl** [in-toto/attestation#98]
In today's [in-toto community call](https://hackmd.io/@lukpueh/ry_e70Qqw) while discussing open ITE's we discussed it being useful for both adopters and maintainers to have a collection of different in-toto attestations that exist in the wild.

Some examples I know of:

* [SLSA provenance (v0.2)](https://slsa.dev/provenance/v0.2) – designed to describe how a subject artefact was produced, with several opaque fields to be defined by buildType: `invocation.parameters`, `invocation.environment` and `buildConfig`
  * [Tekton SLSA provenance](https://github.com/tektoncd/chains/blob/main/docs/intoto.md)
  * [GitLab SLSA provenance](https://docs.gitlab.com/ee/ci/runners/configure_runners.html#attestation-format)
* [Cosign generic attestation](https://github.com/sigstore/cosign/blob/main/specs/COSIGN_PREDICATE_SPEC.md)
* [Cosign vulnerability attestation](https://github.com/sigstore/cosign/blob/main/specs/COSIGN_VULN_ATTESTATION_SPEC.md) – designed for container vulnerability scans, has opaque type for scan results

**Comment by joshuagl** [comment]
I'd be happy to file a PR adding a markdown file to the repo with these links in.

**Comment by joshuagl** [comment]
> * [SLSA provenance (v0.2)](https://slsa.dev/provenance/v0.2) – designed to describe how a subject artefact was produced, with several opaque fields to be defined by buildType: `invocation.parameters`, `invocation.environment` and `buildConfig`
>   
>   * [Tekton SLSA provenance](https://github.com/tektoncd/chains/blob/main/docs/intoto.md)
>   * [GitLab SLSA provenance](https://docs.gitlab.com/ee/ci/runners/configure_runners.html#attestation-format)
>   * [Amber SLSA provenance](https://github.com/project-oak/transparent-release/blob/2cb002d380f23c2fc4f5079d5221c1bcf6c036d1/schema/amber-slsa-buildtype/v1/provenance.json)

Need to review these and see how much they differ from the SLSA provenance format. Is it just the opaque fields, or are there other differences and/or extensions?

**Comment by TomHennen** [comment]
I was thinking it would be nice to have some tool to validate SLSA provenance against the spec.  I've certainly run into cases where people claimed to be doing it but some fields were incorrect.

Are there any current best practices on validating JSON schemas?

**Comment by TomHennen** [comment]
json-schema.org seems to be the thing.  We _could_ define schemas using it and then power some or all of the validation using the associated tooling.  Not sure how folks feel about that?

**Comment by lukpueh** [comment]
> [json-schema.org](https://json-schema.org/) seems to be the thing. We could define schemas using it and then power some or all of the validation using the associated tooling. Not sure how folks feel about that?

I support this! I remember looking into json schema some years ago, when we started developing the in-toto reference implementation. Back then spec/tooling did not seem mature enough. Looks like this has changed.

**Comment by joshuagl** [comment]
I agree it would be great to be able to validate provenance against the spec. I'm not familiar with json-schema.org but it does seem to be popular among colleagues faced with similar problems.

**Comment by TomHennen** [comment]
Here's an example of what the json-schema for in-toto _statements_ would look like:

``` json
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://in-toto.io/Statement/v0.1.schema.json",
  "properties": {
    "_type": {
      "description": "TODO",
      "type": "string",
      "const": "https://in-toto.io/Statement/v0.1"
    },
    "subject": {
      "description": "TODO",
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "name": {
            "type": "string"
          },
          "digest": {
            "type": "object",
            "minProperties": 1
          }
        },
        "required": [
          "name",
          "digest"
        ]
      }
    },
    "predicateType": {
      "description": "TODO",
      "type": "string",
      "format": "uri"
    },
    "predicate": {
      "description": "TODO",
      "type": "object"
    }
  },
  "required": [
    "_type",
    "subject",
    "predicateType"
  ]
}
```

We'd probably want to also define individual schemas for each of the 'official' predicate types.

I'd assume people would want us to have some sort of command line validator.  I'd normally default to golang but it doesn't look like the latest golang implementations support the 2020-12 draft.  Perhaps we should pick [a python one](https://json-schema.org/implementations.html#validator-python)?

Thoughts?

**Comment by gerardborst** [comment]
We from [Argos Notary](https://github.com/argosnotary) would very much like to migrate from an in-toto like structure to using attestations.

I have time and, I think, expertise, so I want to create the schemas and the verification process. But I think the proposed schema isn't detailed enough. Especially the `digest`.

The problem is, I think, that the [digest definition](https://github.com/in-toto/attestation/blob/main/spec/v1/digest_set.md) isn't valid JSON as defined in [json.org](https://www.json.org/json-en.html).

```
{
  "<ALGORITHM_1>": "<HEX/BASE64 VALUE>",
  "<ALGORITHM_2>": "<HEX/BASE64 VALUE>",
  ... 
}
```
A JSON object should consist of static keys with values.

This is also the case in `ResourceDescriptor.

In our project we are using openapi, Jackson, Spring Data, MongoDB, Javascript, React, Axis. All these expect valid JSON and you get in a lot of trouble if it isn't.

**Comment by TomHennen** [comment]
> A JSON object should consist of static keys with values.

Can you provide a reference to where that restriction is indicated?  I don't see it at json.org or in [this copy](https://www.ecma-international.org/wp-content/uploads/ECMA-404_2nd_edition_december_2017.pdf) of the standard.  

Also perhaps this should be tracked as a separate issue?

**Comment by gerardborst** [comment]
Hi Tom,

It is all a bit implicit but if you look under "6 objects" in your reference in the text: " name/value pairs.
A name is a string. A single colon token follows each name, separating the name from the value.".

At least I interpret it as, a name is static and the value variable.

As an example we are searching on the hash in the database and it isn't possible to define an index in mongodb as the `digest` is now defined.

**Comment by gerardborst** [comment]
You are right I will create a separate issue.
